Gtest Unit Testing

glennguy opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->
An implementation of googletest with unit tests for various HLS and DASH manifest scenarios

There is a current test implementation however it is quite inflexible. It relies on large (18KB) precomputed text representations of most attributes of the parsed adaptiveTree, and comparing the whole contents of this with the whole contents of a currently parsed tree. Failures can be caused by differences in attributes that are inconsequential to the feature under test, and on failure you're effectively given a haystack.

This new implementation offers a more usable approach that can quickly pinpoint errors, while aiding new contributors in how things should work.

I've tried to be detailed in the naming of the tests, and sticking to a convention for the test filenames. I wasn't sure if I should add some more comments inside some of the tests to better explain the scenario, please let me know if I should.

As can be seen in the supplied screenshot, out of 21 tests there are currently 6 failing. These are currently known issues. PR #607 fixes 4 of these tests.

There are plently more tests to be added of course but I think this is a good start covering some of the base url/domain/effective url issues of recent times, see #531 .

Running the tests:

Windows 10/Visual Studio:
* Configure build:
 ```
     # inputstream.adaptive and xbmc folders side by side
     mkdir inputstream.adaptive/build
     cd inputstream.adaptive/build
     cmake -T host=x64 -G "Visual Studio 15 2017 Win64" \
    -DADDONS_TO_BUILD=inputstream.adaptive \
    -DCMAKE_BUILD_TYPE=Debug \
    -DADDON_SRC_PREFIX=../..  \
    ../../xbmc/cmake/addons 
```
* Open kodi-addons.sln in VS and build
* Close kodi-addons.sln and open inputstream.adaptive-prefix/src/inputstream.adaptive-build/inputstream.adaptive.sln
* Build the Testinpustream.adaptive project after any changes are made
* Build the RUN_TESTS project to run the test executable (runs the .exe with the correct argument pointing to the manifests folder)

Linux:
* Clone inputstream.adaptive and xbmc into the same folder
* `mkdir inputstream.adaptive/build && cd inputstream.adaptive/build`
* `cmake -DADDONS_TO_BUILD=inputstream.adaptive -DADDON_SRC_PREFIX=../../ ../../xbmc/cmake/addons`
* `make`
* `inputstream.adaptive-prefix/src/inputstream.adaptive-build/src/test/Testinputstream.adaptive ../src/test/manifests`


## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->
For a long time a cycle has been observed of:
* New specific configuration/layout of a manifest starts to be observed in use in the wild
* This variant does not work in the current code
* Changes are made to accommodate
* 1 or more different types/configurations of manifest are now broken

Manual testing is now impractical as there are so many small variations and options (especially for DASH)

## How Has This Been Tested?
Tested on Windows 10/Visual Studio and Ubuntu 18.04

## Screenshots (if appropriate):

![image](https://user-images.githubusercontent.com/11909320/111064653-58720780-8509-11eb-844b-8844963f963c.png)


## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [X] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [X] I have added tests to cover my change
- [ ] All new and existing tests passed (please see above)

glennguy committed:
Refactor for unit testing

Create interface for AESDecrypter to be able to override `RenewLicense` which would use kodi::vfs

glennguy committed:
Remove previous CTest suite

glennguy commented:
@phunkyfish if/when you have any spare time :)

kszaq commented on src/test/TestMain.cpp:
```diff
@@ -0,0 +1,13 @@
+#ifdef _WIN32
+  _putenv_s("DATADIR", args[0].c_str());
+#else
+  setenv("DATADIR", args[0].c_str())
```
To make compiler happy, please change this to:
```
setenv("DATADIR", args[0].c_str(), 1);
```

kszaq commented on CMakeLists.txt:
```diff
@@ -75,6 +76,9 @@ add_subdirectory(lib/mpegts)
 
 add_subdirectory(lib/webm_parser)
 
+enable_testing()
+list(APPEND CMAKE_CTEST_ARGUMENTS "--output-on-failure")
+
```
Please keep these within `if(NOT CMAKE_CROSSCOMPILING)` because compiling tests fails when cross-compiling:
```
if(NOT CMAKE_CROSSCOMPILING)
  enable_testing()
  list(APPEND CMAKE_CTEST_ARGUMENTS "--output-on-failure")
  add_subdirectory(src/test)
endif()
```

phunkyfish replied:
Is there a way to make the tests compile while cross compiling? This would be ideal as otherwise the largest platform would not be tested.

kszaq replied:
I guess these test are not platform-specific but it's definitely possible to fix this, will add another comment in a second.

phunkyfish replied:
Great, thanks

wsnipex replied:
unless you hook up an android and iphone to jenkins, you cannot run tests for cross-compiled platforms

kszaq commented on src/test/CMakeLists.txt:
```diff
@@ -0,0 +1,50 @@
+    ../oscompat.cpp
+    )
+
+target_link_libraries(${BINARY} PUBLIC ${CMAKE_PROJECT_NAME} gtest_main gtest)
```
Please add `${CMAKE_DL_LIBS}` in this line to fix cross-compiling.

kszaq replied:
Please add `${CMAKE_DL_LIBS}` or else this still won't cross-compile. Unless this breaks native builds?

phunkyfish commented:
Thanks for this @glennguy.

Are there fixes planned for the other two test  failures?

We should get this green as soon as this goes in.

phunkyfish commented on src/test/TestMain.cpp:
```diff
@@ -0,0 +1,13 @@
+  setenv("DATADIR", args[0].c_str())
+#endif
+  return RUN_ALL_TESTS();
+}
```
New line

phunkyfish commented on src/test/CMakeLists.txt.in:
```diff
@@ -0,0 +1,15 @@
+  BUILD_COMMAND     ""
+  INSTALL_COMMAND   ""
+  TEST_COMMAND      ""
+)
```
New line

phunkyfish commented on CMakeLists.txt:
```diff
@@ -98,3 +101,4 @@ build_addon(inputstream.adaptive ADP DEPLIBS)
 
 include(CPack)
 
+add_subdirectory(src/test)
```
New line

phunkyfish commented on src/test/TestHelper.h:
```diff
@@ -0,0 +1,39 @@
+  uint64_t mock_time = 10000000L;
+  DASHTestTree();
+  uint64_t GetNowTime() override { return mock_time; }
+};
```
New line

phunkyfish commented on src/test/manifests/hls/1v_master.m3u8:
```diff
@@ -0,0 +1,4 @@
+#EXTM3U
+#EXT-X-VERSION:3
+#EXT-X-STREAM-INF:PROGRAM-ID=1,BANDWIDTH=2119734,CODECS="avc1.77.31, mp4a.40.2",RESOLUTION=960x540
+ts_aes_uriwithslash_chunklist.m3u8
```
New line

phunkyfish commented on src/test/manifests/hls/redirect_absolute_1v_master.m3u8:
```diff
@@ -0,0 +1,5 @@
+#EXT-X-VERSION:3
+#EXT-X-INDEPENDENT-SEGMENTS
+#EXT-X-STREAM-INF:BANDWIDTH=582476,AVERAGE-BANDWIDTH=581432,RESOLUTION=480x270,FRAME-RATE=25.000,CODECS="avc1.42C015,mp4a.40.2"
+https://bit.ly/abcd
```
New line

phunkyfish commented on src/test/manifests/hls/ts_aes_keyuriabsolute_stream_0.m3u8:
```diff
@@ -0,0 +1,18 @@
+#EXTINF:9.033,
+media-abc_84.ts
+#EXTINF:10.0,
+media-abc_85.ts
```
New line

phunkyfish commented on src/test/manifests/hls/ts_aes_keyurirelative_stream_0.m3u8:
```diff
@@ -0,0 +1,18 @@
+#EXTINF:9.033,
+media-abc_84.ts
+#EXTINF:10.0,
+media-abc_85.ts
```
New line

phunkyfish commented on src/test/manifests/hls/ts_aes_keyuriwithslash_stream_0.m3u8:
```diff
@@ -0,0 +1,18 @@
+#EXTINF:9.033,
+media-abc_84.ts
+#EXTINF:10.0,
+media-abc_85.ts
```
New line

phunkyfish commented on src/test/manifests/mpd/segmentbase.mpd:
```diff
@@ -0,0 +1,24 @@
+      <Role value="main" schemeIdUri="urn:mpeg:dash:role:2011"/>
+    </AdaptationSet>
+  </Period>
+</MPD>
```
New line

phunkyfish commented on src/test/manifests/mpd/segtpl.mpd:
```diff
@@ -0,0 +1,17 @@
+      <Representation bandwidth="300000" codecs="avc1.64001e" frameRate="60/2" height="360" id="V300" sar="1:1" width="640" />
+    </AdaptationSet>
+  </Period>
+</MPD>
```
New line

phunkyfish commented on src/test/manifests/mpd/segtpl_baseurl_noslashs.mpd:
```diff
@@ -0,0 +1,23 @@
+      <Representation width="256" height="144" scanType="progressive" id="video=66000" bandwidth="66000"/>
+    </AdaptationSet>
+  </Period>
+</MPD>
```
New line

phunkyfish commented on src/test/manifests/mpd/segtpl_baseurlinmpd.mpd:
```diff
@@ -0,0 +1,18 @@
+      <Representation bandwidth="300000" codecs="avc1.64001e" frameRate="60/2" height="360" id="V300" sar="1:1" width="640" />
+    </AdaptationSet>
+  </Period>
+</MPD>
```
New line

phunkyfish commented on src/test/manifests/mpd/segtpl_noslash_baseurl_slash.mpd:
```diff
@@ -0,0 +1,23 @@
+      <Representation width="256" height="144" scanType="progressive" id="video=66000" bandwidth="66000"/>
+    </AdaptationSet>
+  </Period>
+</MPD>
```
New line

phunkyfish commented on src/test/manifests/mpd/segtpl_slash_baseurl_noslash.mpd:
```diff
@@ -0,0 +1,23 @@
+      <Representation width="256" height="144" scanType="progressive" id="video=66000" bandwidth="66000"/>
+    </AdaptationSet>
+  </Period>
+</MPD>
```
New line

phunkyfish commented on src/test/manifests/mpd/segtpl_slash_baseurl_slash.mpd:
```diff
@@ -0,0 +1,23 @@
+      <Representation width="256" height="144" scanType="progressive" id="video=66000" bandwidth="66000"/>
+    </AdaptationSet>
+  </Period>
+</MPD>
```
New line

phunkyfish reviewed (changes requested):
Can you detail how to see the test results also?

glennguy commented:
Thanks @phunkyfish and @kszaq!

PR #606 fixes one more, but I haven't tried yet for a solution to the remaining case. @matthuisman has done the research on it, I documented in new issue #617. One of us will find a solution and PR for it ASAP.

I'll update the OP with instructions to run tests in the current state

kszaq commented on src/test/TestMain.cpp:
```diff
@@ -0,0 +1,13 @@
+#ifdef _WIN32
+  _putenv_s("DATADIR", args[0].c_str());
+#else
+  setenv("DATADIR", args[0].c_str(), 1)
```
Missing a semicolon here. `;`

glennguy replied:
fudge, sorry. fixed

kszaq commented on src/test/CMakeLists.txt:
```diff
@@ -0,0 +1,50 @@
+    ../oscompat.cpp
+    )
+
+target_link_libraries(${BINARY} PUBLIC ${CMAKE_PROJECT_NAME} gtest_main gtest)
```
Please add `${CMAKE_DL_LIBS}` or else this won't cross-compile. Unless it breaks native builds?

glennguy replied:
sorry my bad, forgot to ctrl+s that change

wsnipex commented on src/test/CMakeLists.txt:
```diff
@@ -0,0 +1,50 @@
+  WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/googletest-download )
+if(result)
+  message(FATAL_ERROR "Build step for googletest failed: ${result}")
+endif()
```
gtest must be added to dependencies, not like this.

glennguy replied:
What do you suggest? I went with the recommended method from [here](https://github.com/google/googletest/blob/master/googletest/README.md#incorporating-into-an-existing-cmake-project) but would it be preferable to copy the source to the lib folder?
I found Kodi uses [this](https://github.com/xbmc/xbmc/blob/master/cmake/modules/FindGtest.cmake), am I better off trying to make use of that?

phunkyfish replied:
There is a system built into add-ons which allows building dependencies. If you look at ffmpegdirect you will see all the dependencies in the `depends` folder. 

I think adaptive has it for expat.

wsnipex replied:
as said, you have to add it to depends: https://github.com/xbmc/inputstream.adaptive/tree/Matrix/depends

smf007 replied:
This comment appears to be still open.  Just tried this on Linux, and cmake completely misses downloading googletest on my system.

glennguy replied:
@smf007 yes I'm working on changing to the same method that all the other depends use. I'll have more time over the next few days to finish but I think I'm 90% there

wsnipex commented on CMakeLists.txt:
```diff
@@ -97,20 +101,4 @@ build_addon(inputstream.adaptive ADP DEPLIBS)
-  add_test(6play_live Test_DashTree "${CMAKE_SOURCE_DIR}/src/parser/testfiles/6play.mpd")
-endif()
\ No newline at end of file
+add_subdirectory(src/test)
```
this won't work for cross-compile, so it has to be
if(NOT CMAKE_CROSSCOMPILING)

glennguy replied:
Sure thing. I saw the comment above about the need to run the compiled tests on Android/ios etc. The tests in this PR are not touching any platform specific code anyway.

glennguy replied:
Done

glennguy commented:
Updated OP to give instructions on running tests - Windows/VS is straight forward but I'll look at making the command in linux a bit more user friendly.

wsnipex commented on FindGTest.cmake:
```diff
@@ -0,0 +1,84 @@
+#
+#   Gtest::Gtest   - The gtest library
+
+if(ENABLE_INTERNAL_GTEST)
```
this whole block is not needed, ENABLE_INTERNAL_GTEST exists only in kodi, not for addons

glennguy replied:
Sure thing, updated.

glennguy committed:
gtest implementation

phunkyfish commented:
And can we run this for Windows/Linux/Mac as part of the Jenkins CI now?

smf007 commented:
I think there might be something still missing, or there is an instruction step missing.  I've tried the latest on Linux and I'm not getting the gtest stuff.

For what it's worth, the gtest stuff in kodi is working fine on my system.  Just not for IA with this pull request.

EDIT :: Just blew away my area and placed this pull in the directory name of "./inputstream.adaptive" instead of "./inputstream.adaptive-new-tests", and it now works.  Sorry for that noise.

glennguy commented:
@phunkyfish No Jenkins CI just yet but I've got a local Jenkins set up to try and work something out.

In the interim so we can get this in and the above mentioned PRs I could modify the current TravisCI script in this repo to run the test command? New PRs and rebased existing PRs could then take advantage.

phunkyfish commented:
> @phunkyfish No Jenkins CI just yet but I've got a local Jenkins set up to try and work something out.
> 
> In the interim so we can get this in and the above mentioned PRs I could modify the current TravisCI script in this repo to run the test command? New PRs and rebased existing PRs could then take advantage.

For the existing PRs that makes tests pass a report from you of those is sufficient. Then we can at least get those in. Once they all pass then we can worry about integrating with Jenkins fully.

glennguy commented:
#606 and #607 look good to me.

We are left with one test failing (issue #617) but a much larger refactor is going to be in order to fix that properly. #620 is aimed at that but I think we'll need a lot more tests added to be sure of that one as it will affect a *lot* of code.

I can put a `GTEST_SKIP()` there for now if you're happy with the plan to address that issue at a later date? At the moment as far as I can tell it's more of a theoretical issue rather than one that's causing any current problem (nothing in any open issues regarding this)

phunkyfish commented:
Ok, let me bring in those two this weekend and we can see where to go from there.

glennguy committed:
Remove TravisCI/Add GitHub Actions

glennguy commented:
Update:
* Added GitHub Actions workflow
* Skipped remaining test that needs a larger rework to pass correctly
* Removed TravisCI conifg (travis-ci.org is due to go offline anytime now)
* Bumped cmake_minimum_required to 3.10 (thanks to @fuzzard for pointing out FindEXPAT is only a part of 3.10+)
* CMakeLists setting of KODI_INCLUDE_DIR is now consistent with other binary addons (append `../` workaround)

phunkyfish reviewed (approved):
LGTM
